[WebVTT] Apply ::cue pseudo to the correct WebVTT object (root)

jernoble opened the pull request:
#### 75cef4fa5fd2155fef875e02a8cee615b78e6af6
<pre>
[WebVTT] Apply ::cue pseudo to the correct WebVTT object (root)
<a href="https://bugs.webkit.org/show_bug.cgi?id=276089">https://bugs.webkit.org/show_bug.cgi?id=276089</a>
<a href="https://rdar.apple.com/130913431">rdar://130913431</a>

Reviewed by Tim Nguyen.

Previously, we applied the ::cue pseudo to the m_cueHighlightBox WebVTT object, in order to
implement the portion of the WebVTT spec where ::cue { background: foo; } applies to the &quot;cue
background box&quot; rather than the &quot;list of WebVTT node objects&quot;. However, this causes other problems,
as it means style in the page cannot override styles applied to the root (i.e., display tree)
WebVTT object.

In order to implement this &quot;exception to CSS styling&quot; for `background:` property, when an incoming
style rule targetting the ident `::cue` selector is added, clone that rule and replace the selector
with a new selector for `[useragentpart=&quot;cue-background&quot;]`, with a separate list of &quot;allowed
properties&quot; containing only `background:`.

Then change all the default styles to be in terms of `::cue` and `[useragentpart=&quot;cue&quot;]`.

* Source/WebCore/Modules/modern-media-controls/controls/text-tracks.css:
([useragentpart=&quot;cue&quot;]):
([useragentpart=&quot;cue&quot;] .hidden):
([useragentpart=&quot;cue&quot;] ruby):
([useragentpart=&quot;cue&quot;] ruby &gt; rt):
(::cue):
(::cue(rt)):
(::cue(i)):
(::cue(b)):
(::cue(u)):
(::-webkit-media-text-track-display): Deleted.
([useragentpart=&quot;-webkit-media-text-track-display&quot;] b): Deleted.
([useragentpart=&quot;-webkit-media-text-track-display&quot;] u): Deleted.
([useragentpart=&quot;-webkit-media-text-track-display&quot;] i): Deleted.
([useragentpart=&quot;-webkit-media-text-track-display&quot;] .hidden): Deleted.
([useragentpart=&quot;-webkit-media-text-track-display&quot;] ruby): Deleted.
([useragentpart=&quot;-webkit-media-text-track-display&quot;] ruby &gt; rt): Deleted.
([useragentpart=&quot;-webkit-media-text-track-display&quot;] ruby &gt; :not(ruby)): Deleted.
* Source/WebCore/css/CSSPseudoSelectors.json:
* Source/WebCore/html/track/VTTCue.cpp:
(WebCore::VTTCue::obtainCSSBoxes):
* Source/WebCore/style/PropertyAllowlist.cpp:
(WebCore::Style::isValidCueStyleProperty):
(WebCore::Style::isValidCueSelectorStyleProperty):
(WebCore::Style::isValidCueBackgroundStyleProperty):
* Source/WebCore/style/PropertyAllowlist.h:
* Source/WebCore/style/PropertyCascade.cpp:
(WebCore::Style::PropertyCascade::addMatch):
* Source/WebCore/style/RuleData.cpp:
(WebCore::Style::determinePropertyAllowlist):
* Source/WebCore/style/RuleSet.cpp:
(WebCore::Style::RuleSet::addRule):

Canonical link: <a href="https://commits.webkit.org/282320@main">https://commits.webkit.org/282320@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d370de6d41b9d4ee1149cc1e9b7644ab52b8bc31

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/60795 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/40154 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/13371 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/64726 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/11342 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/62925 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/47830 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/11617 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/49165 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/7876 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/62829 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/37395 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/52673 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/29996 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/34082 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/9899 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/10255 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/55920 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/10197 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/66455 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/4739 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/10026 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/56531 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/4760 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/52641 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/56726 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/13935 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/3934 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/35959 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/37041 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/38134 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/36786 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/css/CSSPseudoSelectors.json:
```diff
@@ -514,6 +514,11 @@
             "conditional": "ENABLE(VIDEO)",
             "user-agent-part-string": true
         },
+        "cue-background": {
```
If this is only used in UA stylesheet, I would make this change:

```suggestion
        "-internal-cue-background": {
```

Otherwise the web can use it and do feature detection/hacks that can cause web compat problems later on

nt1m commented on Source/WebCore/style/RuleData.cpp:
```diff
@@ -142,8 +142,15 @@ static inline PropertyAllowlist determinePropertyAllowlist(const CSSSelector* se
     for (const CSSSelector* component = selector; component; component = component->tagHistory()) {
 #if ENABLE(VIDEO)
-        if (component->match() == CSSSelector::Match::PseudoElement && (component->pseudoElement() == CSSSelector::PseudoElement::Cue || component->value() == UserAgentParts::cue()))
+        // Property allow-list for `::cue`:
+        if (component->match() == CSSSelector::Match::PseudoElement && component->pseudoElement() == CSSSelector::PseudoElement::UserAgentPart && component->value() == UserAgentParts::cue())
             return PropertyAllowlist::Cue;
+        // Property allow-list for `::cue(selector)`:
+        if (component->match() == CSSSelector::Match::PseudoElement && component->pseudoElement() == CSSSelector::PseudoElement::Cue)
+            return PropertyAllowlist::CueSelector;
```
Why can't those 2 cases use the same allowlist?

nt1m commented on Source/WebCore/Modules/modern-media-controls/controls/text-tracks.css:
```diff
@@ -52,26 +52,36 @@ video[controls]::-webkit-media-text-track-container.visible-controls-bar {
     writing-mode: writing-mode;
     overflow-wrap: break-word;
-    white-space: pre-line;
     text-wrap: balance;
```
text-wrap: balance seems allowed via ::cue.

jernoble committed:
[WebVTT] Apply ::cue pseudo to the correct WebVTT object (root)
https://bugs.webkit.org/show_bug.cgi?id=276089
rdar://130913431

Reviewed by Tim Nguyen.

Previously, we applied the ::cue pseudo to the m_cueHighlightBox WebVTT object, in order to
implement the portion of the WebVTT spec where ::cue { background: foo; } applies to the "cue
background box" rather than the "list of WebVTT node objects". However, this causes other problems,
as it means style in the page cannot override styles applied to the root (i.e., display tree)
WebVTT object.

In order to implement this "exception to CSS styling" for `background:` property, when an incoming
style rule targetting the ident `::cue` selector is added, clone that rule and replace the selector
with a new selector for `[useragentpart="cue-background"]`, with a separate list of "allowed
properties" containing only `background:`.

Then change all the default styles to be in terms of `::cue` and `[useragentpart="cue"]`.

* Source/WebCore/Modules/modern-media-controls/controls/text-tracks.css:
([useragentpart="cue"]):
([useragentpart="cue"] .hidden):
([useragentpart="cue"] ruby):
([useragentpart="cue"] ruby > rt):
(::cue):
(::cue(rt)):
(::cue(i)):
(::cue(b)):
(::cue(u)):
(::-webkit-media-text-track-display): Deleted.
([useragentpart="-webkit-media-text-track-display"] b): Deleted.
([useragentpart="-webkit-media-text-track-display"] u): Deleted.
([useragentpart="-webkit-media-text-track-display"] i): Deleted.
([useragentpart="-webkit-media-text-track-display"] .hidden): Deleted.
([useragentpart="-webkit-media-text-track-display"] ruby): Deleted.
([useragentpart="-webkit-media-text-track-display"] ruby > rt): Deleted.
([useragentpart="-webkit-media-text-track-display"] ruby > :not(ruby)): Deleted.
* Source/WebCore/css/CSSPseudoSelectors.json:
* Source/WebCore/html/track/VTTCue.cpp:
(WebCore::VTTCue::obtainCSSBoxes):
* Source/WebCore/style/PropertyAllowlist.cpp:
(WebCore::Style::isValidCueStyleProperty):
(WebCore::Style::isValidCueSelectorStyleProperty):
(WebCore::Style::isValidCueBackgroundStyleProperty):
* Source/WebCore/style/PropertyAllowlist.h:
* Source/WebCore/style/PropertyCascade.cpp:
(WebCore::Style::PropertyCascade::addMatch):
* Source/WebCore/style/RuleData.cpp:
(WebCore::Style::determinePropertyAllowlist):
* Source/WebCore/style/RuleSet.cpp:
(WebCore::Style::RuleSet::addRule):

Canonical link: https://commits.webkit.org/282320@main
